### DISCUSSION TITLE
How to avoid class duplication when base class ctor denotes nullable param?

### DISCUSSION THREAD

**Discussion by deep-outcome** [dotnet/roslyn#58448] [Ideas]
When generic class is defined, its type parameter (`T`) is _considered_ to be a class. I mean

````csharp
class C<T>
{
  T? nt;
  C ( T? nt ) => this.nt = nt;
}

private hidebysig specialname rtspecialname 
		instance void .ctor (
			!T nt
		) cil managed 

class C2<T> where T : class
{
  T? nt;
  C2 ( T? nt ) => this.nt = nt;
}

private hidebysig specialname rtspecialname 
		instance void .ctor (
			!T nt
		) cil managed 
```` 

while

```` csharp
class C1<T> where T : struct
{
  T? nt;
  C1 ( T? nt ) => this.nt = nt;
}

private hidebysig specialname rtspecialname 
		instance void .ctor (
			valuetype [System.Runtime]System.Nullable`1<!T> nt
		)
```` 

So this in turn with fact that it is not allowed to specify constraint (`where T : xxx`) on .ctors disallows this

````csharp
class C<T>
{
  public T? NT { get; }
  public C ( T? nt ) => NT = nt;
}

class D<T> : C<T>
where T : struct
{
  D ( T? nt )
  : base (nt) //Error CS1503 Argument 1: cannot convert from 'T?' to 'T'
  { }
}
```` 

In other words generic type parameter constraint is type specific so derived class cannot uplift its constraint to base.

So all above means there has to be 2 `C`: `class C<T> where T : struct` and `class C<T> where T : class`. For small models this is not a problem but handling 2 big models can render quite odd.

I guess allowing `struct` and `class` generic constraints on .ctor can be fine. A I missing something?

(This declaration generates many errors with syntax, duplicities, etc.)
````csharp
class C<T>
{
  public T? NT { get; }
  public C ( T? nt ) where T: struct => NT = nt;
  public C ( T? nt ) where T: class => NT = nt;
}
````

**Comment by CyrusNajmabadi** [comment]
> When generic class is defined, its type parameter (T) is considered to be a class. I mean

This is not the case.  T is neither a struct nor a class.  And `T?` simply means "this value is `default`able" (as opposed to `nullable`).

> So this in turn with fact that it is not allowed to specify constraint (where T : xxx) on .ctors disallows this

True.  but you are allowed to write:

```c#
class C<T>
{
    public T? NT { get; }
    public C(T? nt) => NT = nt;
}

class D<T> : C<T> where T : struct
{
    D(T nt) : base(nt)
    {
    }
}
```

> In other words generic type parameter constraint is type specific so derived class cannot uplift its constraint to base.

I don't really know what this means.

**Reply by deep-outcome** [reply]
> This is not the case. T is neither a struct nor a class. And T? simply means "this value is defaultable" (as opposed to nullable).

Yes, this is truth. I was just about to show that `T?` is never considered `Nullable<T>` unless explicitly specified.

> True. but you are allowed to write:

Yes, but not `new D<int> (null);`.

Clue behind this is something like

````csharp
class Base<T>
{
  public T MustBe { get; }
  public T? MayBe { get; }
  public Base
  (
    T mustBe,
    T? mayBe
  )
  {
    MustBe = mustBe;
    MayBe = mayBe;
  }
}
````

Where `T? mayBe` provides some information based on context. There can be `default` but to carry on information whether `default` is that `default` or value that equals `default` one needs additional info – like `bool mayBeIsValue` basically _decomposition_ of `Nullable<T>`.

> I don't really know what this means.

I mean **generic type parameter constraint** is applied at type level same as **generic type parameter** that means constraint is bounded to type definition (it is type specific). Any derived class cannot _override_ base class **generic type parameter constraint** because it would then override base type **type definition**.

> I guess allowing struct and class generic constraints on .ctor can be fine. A I missing something?

On other hand it demands some facet around 2 types that would be then represented by 1 class/struct.

**Reply by deep-outcome** [reply]
I guess I found my answer. There is no simple built-in solution to this. Type must be duplicated in order to provide both nullable variants.
